### PR TITLE
*: convert lockKeys to key flag

### PR DESCRIPTION
**Pull request by bobotu**
### What problem does this PR solve?
* Convert `tikvTxn.lockKeys` and `tikvTxn.lockedMap` to `KeyFlags` in `MemBuffer`.
* Simplify `twoPhaseCommitter.initKeysAndMutations`.
* Fix some typo in `memdb.go`.

### Check List 

Tests 
- Unit test

### Release note 

-  Reduce large pessimistic transaction's memory usage.

### CODE REVIEW

**Comment by tiancaiamao** [comment]
LGTM

**Comment by bobotu** [comment]
PTAL @cfzjywxk

File: sessionctx/stmtctx/stmtctx.go [hunk 1]
```diff
@@ -486,7 +485,6 @@ func (sc *StatementContext) ResetForRetry() {
 	sc.TableIDs = sc.TableIDs[:0]
 	sc.IndexNames = sc.IndexNames[:0]
 	sc.TaskID = AllocateTaskID()
-	sc.CheckKeyExists = make(map[string]struct{})
```
Comment (cfzjywxk) [inline sessionctx/stmtctx/stmtctx.go:489]: We need to reset check exists flags on the keys if the statement execution has failed and statement will be retried, like the pessimistic write conflict retry.

File: store/tikv/txn.go [hunk 2]
```diff
@@ -431,19 +451,19 @@ func (txn *tikvTxn) LockKeys(ctx context.Context, lockCtx *kv.LockCtx, keysInput
...
 			val, _ := lockCtx.Values[string(key)]
-			valExists := len(val.Value) > 0
-			txn.lockedMap[string(key)] = valExists
-		} else {
-			txn.lockedMap[string(key)] = true
+			if len(val.Value) == 0 {
+				valExists = kv.SetKeyLockedValueNotExists
+			}
 		}
+		memBuf.UpdateFlags(key, kv.SetKeyLocked, kv.DelNeedCheckExists, valExists)
```
Comment (cfzjywxk) [inline store/tikv/txn.go:464]: Maybe we need to reset some already locked keys which have check exist needs. It's statement level and maybe there will be some flags left?

File: sessionctx/stmtctx/stmtctx.go (hunk 1 above)
Comment (bobotu) [reply sessionctx/stmtctx/stmtctx.go:489]: `handlePessimisticLockError` will rollback all changes made in the current statement by `StmtRollback`. `KeyFlags` will be discarded in this step.

File: store/tikv/txn.go (hunk 2 above)
Comment (bobotu) [reply store/tikv/txn.go:464]: I have done three flag changes in this `UpdateFlags`:
1. Mark key is locked
2. Delete need check exists flag
3. Record the existence of value

**Comment by cfzjywxk** [comment]
LGTM

**Pull request merged**